core/src/main/java/com/datastrato/gravitino/catalog/CapabilityHelpers.java

```diff
@@ -25,22 +35,212 @@ public static TableChange[] applyCapabilities(Capability capabilities, TableChan
+    Namespace namespace = ident.namespace();
+
+    String sensitiveOnMetalakeName = applyCaseSensitiveOnMetalakeName(namespace.level(0));
+    namespace.levels()[0] = sensitiveOnMetalakeName;
+    applyNameSpecificationOnMetalake(namespace.level(0));
+
+    String sensitiveOnCatalogName = applyCaseSensitiveOnCatalogName(namespace.level(1));
+    namespace.levels()[1] = sensitiveOnCatalogName;
+    applyNameSpecificationOnCatalog(namespace.level(1));
+
+    switch (scope) {
+      case SCHEMA:
+        break;
+      case FILESET:
+      case TOPIC:
+      case TABLE:
+        String sensitiveOnSchemaName =
+            applyCaseSensitiveOnName(SCHEMA, namespace.level(namespace.length() - 1), capabilities);
+        namespace.levels()[namespace.length() - 1] = sensitiveOnSchemaName;
+        applyNameSpecification(SCHEMA, sensitiveOnSchemaName, capabilities);
+        break;
+      default:
+        throw new IllegalArgumentException("Unexpected capability scope: " + scope);
+    }
+    String name = applyCaseSensitiveOnName(scope, ident.name(), capabilities);
+    applyNameSpecification(scope, name, capabilities);
+    return NameIdentifier.of(namespace, name);
+  }
+
+  public static Transform[] applyCapabilities(Transform[] transforms, Capability capabilities) {
+    return Arrays.stream(transforms)
+        .map(t -> applyCapabilities(t, capabilities))
+        .toArray(Transform[]::new);
+  }
+
+  public static Distribution applyCapabilities(Distribution distribution, Capability capabilities) {
+    applyCapabilities(distribution.expressions(), capabilities);
+    return distribution;
+  }
+
+  public static SortOrder[] applyCapabilities(SortOrder[] sortOrders, Capability capabilities) {
+    return Arrays.stream(sortOrders)
+        .map(s -> applyCapabilities(s, capabilities))
+        .toArray(SortOrder[]::new);
+  }
+
+  public static Index[] applyCapabilities(Index[] indexes, Capability capabilities) {
+    return Arrays.stream(indexes)
+        .map(i -> applyCapabilities(i, capabilities))
+        .toArray(Index[]::new);
+  }
+
+  public static NameIdentifier applyCapabilitiesOnMetalake(NameIdentifier ident) {
+    String sensitiveOnMetalakeName = applyCaseSensitiveOnMetalakeName(ident.name());
+    applyNameSpecificationOnMetalake(sensitiveOnMetalakeName);
+
+    return NameIdentifier.of(ident.namespace(), sensitiveOnMetalakeName);
+  }
+
+  public static NameIdentifier applyCapabilitiesOnCatalog(NameIdentifier ident) {
+    Namespace namespace = ident.namespace();
+
+    String sensitiveOnMetalakeName = applyCaseSensitiveOnMetalakeName(namespace.level(0));
+    namespace.levels()[0] = sensitiveOnMetalakeName;
+    applyNameSpecificationOnMetalake(namespace.level(0));
+
+    String sensitiveOnCatalogName = applyCaseSensitiveOnCatalogName(ident.name());
+    applyNameSpecificationOnCatalog(ident.name());
+    return NameIdentifier.of(namespace, sensitiveOnCatalogName);
+  }
```

Comment:
I would suggest remove the metalake and catalog related check in this PR. I think it is out of the scope of the Capability framework, we can do it separately.

Reply:
removed, please help review again, thx